multicluster/controllers/multicluster/label_identity_export_controller.go

```diff
@@ -0,0 +1,296 @@
+			addedLabels = append(addedLabels, l)
+		}
+	}
+	for l := range deletedClusterLabels {
+		deletedCluster := sets.NewString(clusterID)
+		if clusters, ok := r.labelsToClusters[l]; ok && clusters.Equal(deletedCluster) {
+			// The cluster where the label identity is being deleted was the only cluster that has
+			// the label identity. Hence, the label identity is no longer present in the ClusterSet.
+			deletedLabels = append(deletedLabels, l)
+		}
+	}
+	if len(addedLabels)+len(deletedLabels) == 0 {
+		return nil
+	}
+	reconcileFailedLabels := sets.NewString()
+	r.handleLabelIdentitiesDelete(ctx, deletedLabels, reconcileFailedLabels, clusterID, resExport)
+	r.handleLabelIdentitiesAdd(ctx, addedLabels, reconcileFailedLabels, clusterID, resExport)
+	if len(reconcileFailedLabels) > 0 {
+		return fmt.Errorf("failed to reconcile LabelIdentityImport for labels %v", reconcileFailedLabels)
+	}
+	return nil
+}
+
+// handleLabelIdentitiesDelete deletes LabelIdentityImports of label identities that no longer exists
+// in the ClusterSet. Note that the ID of a label identity is only released if the deletion of its
+// LabelIdentityImport succeeded.
+func (r *LabelIdentityExportReconciler) handleLabelIdentitiesDelete(ctx context.Context,
+	deletedLabels []string, reconcileFailedLabels sets.String, clusterID string, resExport *mcsv1alpha1.ResourceExport) {
+	for _, label := range deletedLabels {
+		existingLabelImport := &mcsv1alpha1.LabelIdentityImport{}
+		resNamespaced := types.NamespacedName{
+			Name:      hashLabelIdentity(label),
+			Namespace: resExport.Namespace,
+		}
+		err := r.Client.Get(ctx, resNamespaced, existingLabelImport)
+		if err != nil && !apierrors.IsNotFound(err) {
+			klog.ErrorS(err, "failed to get LabelIdentityImport for stale label", "label", label)
+			reconcileFailedLabels.Insert(label)
+			continue
+		}
+		if err := r.Client.Delete(ctx, existingLabelImport, &client.DeleteOptions{}); err != nil {
+			klog.ErrorS(err, "failed to delete LabelIdentityImport for stale label", "label", label)
+			reconcileFailedLabels.Insert(label)
+			continue
+		}
+		// Delete caches for the label and release the ID assigned for the label
+		id, _ := r.labelsToID[label]
+		r.allocator.release(id)
+		delete(r.labelsToID, label)
+		delete(r.labelsToClusters, label)
+		if clusterLabels, ok := r.clusterToLabels[clusterID]; ok {
+			clusterLabels.Delete(label)
+		}
+	}
+}
+
+// handleLabelIdentitiesDelete creates LabelIdentityImports of label identities that are added
+// in the ClusterSet. Note that the ID of a label identity is only allocated and stored if the
+// creation of its LabelIdentityImport succeeded.
+func (r *LabelIdentityExportReconciler) handleLabelIdentitiesAdd(ctx context.Context,
+	addedLabels []string, reconcileFailedLabels sets.String, clusterID string, resExport *mcsv1alpha1.ResourceExport) {
+	for _, label := range addedLabels {
+		id, err := r.allocator.allocate()
+		if err != nil {
```

Comment:
I suppose you can return immediately if ID allocate failed? since you only have one error which is `no ID available`.

Reply:
`reconcileFailedLabels` won't contain all the failed to reconcile labels in that case. But since the only usage now is to check if `len(reconcileFailedLabels) > 0`, I suppose it's okay.